### PR TITLE
fix activation collection and add gradient pruners

### PR DESCRIPTION
**Pull request by colorjam**
1. Initialization of collected_activation point to the same memory
2. Add filter pruner based on Taylor expansion in [2019CVPR] Importance Estimation for Neural Network Pruning

### CODE REVIEW

File: src/sdk/pynni/nni/compression/torch/activation_rank_filter_pruners.py [hunk 1]
```diff
@@ -35,10 +35,13 @@ def __init__(self, model, config_list, optimizer=None, activation='relu', statis
 
         super().__init__(model, config_list, optimizer)
         self.set_wrappers_attribute("if_calculated", False)
-        self.set_wrappers_attribute("collected_activation", [])
+        self.set_wrappers_attribute("collected_activation", None)
```
Comment (QuanluZhang) [inline src/sdk/pynni/nni/compression/torch/activation_rank_filter_pruners.py]: we should do deep copy within `set_wrappers_attribute` or add document to tell users if passing an object, the object will be shared by all the wrappers. @chicm-ms what do you think?

File: src/sdk/pynni/nni/compression/torch/gradient_rank_filter_pruners.py [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+        """
+        Calculate the mask of given layer.
+        Filters with the smallest importance criterion which is calculated from the activation are masked.
+
+        Parameters
+        ----------
+        layer : LayerInfo
+            the layer to instrument the compression operation
+        config : dict
+            layer's pruning config
```
Comment (QuanluZhang) [inline src/sdk/pynni/nni/compression/torch/gradient_rank_filter_pruners.py]: please update docstring here, i know you copied it here, which means some docstrings somewhere else have not been updated...

File: src/sdk/pynni/nni/compression/torch/gradient_rank_filter_pruners.py [hunk 3]
```diff
@@ -0,0 +1,156 @@
+# Copyright (c) Microsoft Corporation.
```
Comment (QuanluZhang) [inline src/sdk/pynni/nni/compression/torch/gradient_rank_filter_pruners.py]: please update the corresponding doc files for this pruner under `https://github.com/microsoft/nni/tree/master/docs/en_US/Compressor`.

File: src/sdk/pynni/nni/compression/torch/activation_rank_filter_pruners.py (hunk 1 above)
Comment (chicm-ms) [reply src/sdk/pynni/nni/compression/torch/activation_rank_filter_pruners.py]: I think deep copy is too much, maybe we can consider to remove `set_wrappers_attribute` if it is too tricky to use it. Both `if_calculated` and `collected_activation` can be implemented without `set_wrappers_attribute`.

File: src/sdk/pynni/nni/compression/torch/utils.py [hunk 4]
```diff
@@ -0,0 +1,280 @@
...
+import numpy as np
+
+from os import path as osp
+from collections import OrderedDict
+
+import pandas as pd
+import torch
+import datetime
+
+class AverageMeter:
```
Comment (chicm-ms) [inline src/sdk/pynni/nni/compression/torch/utils.py]: We should not put these utility classes/functions not related to model compression into model compression SDK

File: docs/zh_CN/Compressor/Pruner.md [hunk 5]
```diff
@@ -14,6 +14,12 @@ NNI Compressor 中的 Pruner
     * [APoZ Rank Pruner](#activationapozrankfilterpruner)
     * [Activation Mean Rank Pruner](#activationmeanrankfilterpruner)
 
+* [具有梯度等级的 Filter Pruners](#gradientfilterpruners)
```
Comment (chicm-ms) [inline docs/zh_CN/Compressor/Pruner.md]: We do not need to update Chinese doc, translation will be done in bunch for each release.

File: examples/model_compress/main.py [hunk 6]
```diff
@@ -0,0 +1,291 @@
+import os
+import argparse
```
Comment (chicm-ms) [inline examples/model_compress/main.py]: Perhaps use another PR to submit examples if they are not specific to gradient pruners?

File: src/sdk/pynni/nni/compression/torch/gradient_rank_filter_pruners.py [hunk 7]
```diff
@@ -0,0 +1,160 @@
...
+        self.old_step = self.optimizer.step
+        self.patch_optimizer(self.calc_contributions)
+
+    def calc_contributions(self):
+        raise NotImplementedError('{} calc_contributions is not implemented'.format(self.__class__.__name__))
+
+    def get_mask(self, base_mask, contribution, num_prune):
+        raise NotImplementedError('{} get_mask is not implemented'.format(self.__class__.__name__))
+
+    def calc_mask(self, wrapper, **kwargs):
```
Comment (chicm-ms) [inline src/sdk/pynni/nni/compression/torch/gradient_rank_filter_pruners.py]: Please add an unit test case in `src/sdk/pynni/tests/test_compressor.py` to test the correctness of calc_mask

File: docs/en_US/Compressor/Pruner.md [hunk 8]
```diff
@@ -11,8 +11,10 @@ Index of supported pruning algorithms
     * [L1Filter Pruner](#l1filter-pruner)
     * [L2Filter Pruner](#l2filter-pruner)
 * [Filter Pruners with Activation Rank](#activationrankfilterpruner)
-    * [APoZ Rank Pruner](#activationapozrankfilterpruner)
-    * [Activation Mean Rank Pruner](#activationmeanrankfilterpruner)
+  * [APoZ Rank Pruner](#activationapozrankfilterpruner)
+  * [Activation Mean Rank Pruner](#activationmeanrankfilterpruner)
```
Comment (QuanluZhang) [inline docs/en_US/Compressor/Pruner.md]: incorrect indent?

File: docs/en_US/Compressor/Pruner.md [hunk 9]
```diff
@@ -341,4 +343,42 @@ You can view example for more information
...
+from nni.compression.torch import TaylorFOWeightFilterPruner
+config_list = [{
+    'sparsity': 0.5,
+    'op_types': ['Conv2d']
+}]
+pruner = TaylorFOWeightFilterPruner(model, config_list)
+pruner.compress()
+```
+
+You can view example for more information
```
Comment (QuanluZhang) [inline docs/en_US/Compressor/Pruner.md]: please add the reference link of this example

File: docs/zh_CN/Compressor/Pruner.md (hunk 5 above)
Comment (QuanluZhang) [reply docs/zh_CN/Compressor/Pruner.md]: yes. @squirrelsc this is the translated one.

File: src/sdk/pynni/nni/compression/torch/activation_rank_filter_pruners.py (hunk 1 above)
Comment (QuanluZhang) [reply src/sdk/pynni/nni/compression/torch/activation_rank_filter_pruners.py]: @chicm-ms It does not solve the problem, I think we should add docstring for API to make it clear

File: docs/en_US/Compressor/Pruner.md [hunk 10]
```diff
@@ -341,4 +343,42 @@ You can view example for more information
...
+
+PyTorch code
+
+```python
+from nni.compression.torch import TaylorFOWeightFilterPruner
+config_list = [{
+    'sparsity': 0.5,
+    'op_types': ['Conv2d']
+}]
+pruner = TaylorFOWeightFilterPruner(model, config_list)
```
Comment (QuanluZhang) [inline docs/en_US/Compressor/Pruner.md]: no optimizer as input argument?

File: docs/en_US/Compressor/Pruner.md (hunk 9 above)
Comment (colorjam) [reply docs/en_US/Compressor/Pruner.md]: examples will add in another PR
